assets/js/googlesitekit/data/create-settings-store.test.js

```diff
@@ -0,0 +1,453 @@
+				} ).toThrow( 'values is required.' );
+			} );
+
+			it( 'sets isDoingSaveSettings', () => {
+				fetch
+					.doMockOnceIf(
+						/^\/google-site-kit\/v1\/core\/site\/data\/settings/
+					)
+					.mockResponseOnce(
+						JSON.stringify( { setting1: true } ),
+						{ status: 200 }
+					);
+
+				dispatch.fetchSaveSettings( {} );
+				expect( select.isDoingSaveSettings() ).toEqual( true );
+			} );
+		} );
+
+		describe( 'receiveSaveSettings', () => {
+			it( 'requires the values param', () => {
+				expect( () => {
+					dispatch.receiveSaveSettings();
+				} ).toThrow( 'values is required.' );
+			} );
+
+			it( 'receives and sets values', () => {
+				const serverValues = { setting1: 'serverside', setting2: 'serverside' };
+				const clientValues = { setting1: 'clientside', setting3: 'clientside' };
+
+				dispatch.setSettings( clientValues );
+				dispatch.receiveSaveSettings( serverValues );
+
+				// Client values are ignored here, server values replace them.
+				expect( store.getState().settings ).toMatchObject( { ...serverValues } );
+			} );
+		} );
+
+		// Tests for "pseudo-action" setSetting, available via setting-specific "set{SettingSlug}".
+		describe( 'setSetting', () => {
+			it( 'has the correct action name', () => {
+				expect( Object.keys( storeDefinition.actions ) ).toEqual(
+					expect.arrayContaining( [ SETTING.action ] )
+				);
+			} );
+
+			it( 'returns the correct action type', () => {
+				const action = storeDefinition.actions[ SETTING.action ]( true );
+				expect( action.type ).toEqual( SETTING.actionType );
+			} );
+
+			it( 'requires the value param', () => {
+				expect( () => {
+					dispatch[ SETTING.action ]();
+				} ).toThrow( 'value is required.' );
+			} );
+
+			it( 'updates the respective setting', () => {
+				const value = 'new';
+
+				dispatch[ SETTING.action ]( value );
+				expect( store.getState().settings ).toMatchObject( { [ SETTING.slug ]: value } );
+			} );
+		} );
+	} );
+
+	describe( 'selectors', () => {
+		describe( 'getSettings', () => {
+			it( 'uses a resolver to make a network request', async () => {
+				const response = { setting1: 'value' };
+				fetch
+					.doMockOnceIf(
+						/^\/google-site-kit\/v1\/core\/site\/data\/settings/
+					)
+					.mockResponseOnce(
+						JSON.stringify( response ),
+						{ status: 200 }
+					);
+
+				const initialSettings = select.getSettings();
+				// Settings will be their initial value while being fetched.
+				expect( initialSettings ).toEqual( undefined );
+				await subscribeUntil( registry,
+					() => (
+						select.getSettings() !== undefined
+					),
+				);
+
+				const settings = select.getSettings();
+
+				expect( fetch ).toHaveBeenCalledTimes( 1 );
+				expect( settings ).toEqual( response );
+
+				expect( fetch ).toHaveBeenCalledTimes( 1 );
+				expect( select.getSettings() ).toEqual( settings );
+			} );
+
+			it( 'returns client settings even if server settings have not loaded', () => {
```

Comment:
If settings from the server simply haven't been loaded yet (and we know that… and we overwrite client settings after fetching from the server)… why not just have this also return `undefined`? It seems odd to let the client save settings we'll never use.

Reply:
We don't actually override client side settings when they're loaded from the server (see `receiveSettings`). We only override client side settings when they were just saved (see `receiveSaveSettings`).

Reasoning: When you save, the UI should be blocked, so we should safely assume server-side settings are in sync with client-side. When the server-side settings are being loaded though, depending on the implementation I'd say it may be acceptable here to already allow for user input.